### PR TITLE
Restructure credentials after environment overrides.

### PR DESCRIPTION
**Pull request by kaspth**
Follow up to: e0d3313

cc @morgoth, sorry I didn't catch these things in your original PR! What do you think about these? 😄
cc @dhh, not sure if you're interested in looking too. It's just that implementation review you wanted …a little later 😁

- Revert renames from `encrypted` and `encrypted_file` back to `credentials`.
  They might be using our Encrypted* generators but from that level of abstraction
  they're still about credentials.

- Same vein: extract a `credentials` method for the `encrypted` local variable. But
  don't call it `encrypted` just because it uses that under the hood. It's about
  capturing the credentials. It's also useful in `change_credentials_in_system_editor`.

- Remove lots of needless argument passing. We've abstracted content_path and key_path
  into methods for a reason, so they should be used. Also spares a conspicuous rename
  of content_path into file_path in other methods.

- Reorders private methods so they're grouped into: command building blocks, option
  parsers, and the generators.

- Extracts commonality in the credentials application tests. A tad unsure about this.
  But I do like that we go with key, content thus matching the command and remove the
  yield which isn't really needed.

- Moves test/credentials_test.rb to beneath the test/application directory. It's a
  Rails application test, so it should be in there.

- Uses `root.join` — a neat trick gleaned from the tests! — and composes the configuration
  private methods such that the building block is below the callers.

### CODE REVIEW

File: railties/lib/rails/commands/credentials/USAGE [hunk 1]
```diff
@@ -41,9 +41,18 @@ from leaking.
...
-To edit given file use command `rails credentials:edit --environment production`
-Default paths can be overwritten by setting `config.credentials.content_path` and `config.credentials.key_path`.
+The `credentials` command supports passing an `--environment` option to create an
+environment specific override. That override will takes precedence over the
+global `config/credentials.yml.enc` file when running in that environment. So:
+
+   rails credentials:edit --environment development
+
+will create `config/credentials/development.yml.enc` with the corresponding
+encryption key in `config/credentials/production.key` if the credentials file
```
Comment (morgoth) [inline railties/lib/rails/commands/credentials/USAGE]: this should say `config/credentials/development.key`

File: railties/lib/rails/application/configuration.rb [hunk 2]
```diff
@@ -278,25 +278,25 @@ def respond_to_missing?(symbol, *)
...
 
       private
-        def credentials_available_for_current_env?
-          File.exist?("#{root}/config/credentials/#{Rails.env}.yml.enc")
-        end
-
         def default_credentials_content_path
           if credentials_available_for_current_env?
-            File.join(root, "config", "credentials", "#{Rails.env}.yml.enc")
+            root.join("config", "credentials", "#{Rails.env}.yml.enc")
```
Comment (morgoth) [inline railties/lib/rails/application/configuration.rb]: This was failing in one test for me as `root` was a string there (not sure if it was a valid usecase). I guess Travis will show it.

**Review by morgoth — APPROVED** [review]

File: railties/CHANGELOG.md [hunk 3]
```diff
@@ -5,12 +5,19 @@
...
+*   Support environment specific credentials overrides.
 
-    For `production` environment look first for `config/credentials/production.yml.enc` file that can be decrypted by
-    `ENV["RAILS_MASTER_KEY"]` or `config/credentials/production.key` master key.
-    Edit given environment credentials file by command `rails credentials:edit --environment production`.
-    Default paths can be overwritten by setting `config.credentials.content_path` and `config.credentials.key_path`.
+    So any environment will look for `config/credentials/#{Rails.env}.yml.enc` and fall back
+    to `config/credentials.yml.enc`.
+
+    The encryption key can be in `ENV["RAILS_MASTER_KEY"]`, `config/credentials/production.key` or `config/credentials/master.key`.
```
Comment (oniofchaos) [inline railties/CHANGELOG.md]: Would `config/credentials/development.yml.enc` and `config/credentials.yml.enc` share the same master key? It seems like this is answered in the `USAGE` docs below but I'm honestly still not 100% sure. Could def use some clarification, and certainly could use highlighting of whether it's the same key vs different keys in the changelog IMO.

**Review by dillonwelch — COMMENTED** [review]
Neat! First time I've seen this so just some feedback based on fresh eyeballs on the feature.

File: railties/lib/rails/application/configuration.rb (hunk 2 above)
Comment (kaspth) [reply railties/lib/rails/application/configuration.rb]: Looks like calling `app.root=` auto converts to a Pathname: https://github.com/rails/rails/blob/7c89948c416fbc32b59e33a0ab454545b4f6fed7/railties/lib/rails/engine/configuration.rb#L74
And the method that initialises the configuration also gets a Pathname from `find_root`: https://github.com/rails/rails/blob/3631d7eee4bd034f2eefe1b9892d5fcd565579ac/railties/lib/rails/application.rb#L379

I think that test is off.

File: railties/CHANGELOG.md (hunk 3 above)
Comment (kaspth) [reply railties/CHANGELOG.md]: Good point! We actually don't fallback to the `master.key` file, which I think I thought back when doing this editing.

**Pull request merged**